RetroAchievements/RAWeb
# feat: add BuildGameListAction

This PR adds a new `BuildGameListAction` class in the `App\Platform\Actions` namespace.

This action is designed to be idempotent and provides context-agnostic game list records, ideal for use in Laravel controllers, such as those constructing Inertia.js reviews for a React.js datatable.

Key features of the action include:
* Support for multiple contexts (eg: Want to Play Games, Want to Develop Games, All Games, System Games, Hubs, and Developed Sets lists). Currently, only the Want to Play Games context is implemented.
* Sorting by any field in each record.
* Filtering by various fields, with support for applying multiple filters simultaneously.
* Reporting the user's progress and awards for each game in the list.
* Performance efficiency, with execution times ranging between 40-70ms under most circumstances, with potential for further optimization.
* Automatically reporting if a game is in the user's Want to Play Games list.
* Elimination of N+1 query issues.
* Nearly 100% statement test coverage.
* End-to-end type safety, with results fed into a `PaginatedData` DTO, similar to how the forum Recent Posts page functions.

**How to test this PR:**
In addition to comprehensive PHPUnit test coverage, a "playground" page has been set up for testing the action.

This page can be visited at http://localhost:64000/demo/playground.

In _playground.blade.php_, you'll find:
```php
render(function (View $view) {
    $user = User::firstWhere("User", "Jamiras");
    $action = new BuildGameListAction();

    $result = $action->execute(
        GameListType::UserPlay,
        $user,
        perPage: 50,
        sort: ["field" => "title", "direction" => "asc"], // see $validSortFields in BuildGameListAction for possible sorts
        filters: [], // eg: ['system' => [1]] or ['system' => [1, 5]] or ['award' => ['completed', 'mastered']]
    );

    return $view->with(['result' => $result]);
});
```

You can use this playground to experiment with various parameters of the action.

Currently, the playground page is the only place in RAWeb's UI where this action is exposed. This PR aims to get the back-end logic reviewed. A separate front-end PR will follow to implement the React.js datatable that will consume the action results. Splitting the PRs this way ensures logical separation between back-end and front-end components.
